iroh-net/src/discovery/local_swarm_discovery.rs

```diff
@@ -0,0 +1,342 @@
+            }
+        });
+        Ok(Self {
+            handle,
+            sender: send,
+        })
+    }
+
+    fn spawn_discoverer(
+        node_id: PublicKey,
+        sender: Sender<Message>,
+        socketaddrs: BTreeSet<SocketAddr>,
+        rt: &tokio::runtime::Handle,
+    ) -> Result<DropGuard> {
+        let callback = move |node_id: &str, peer: &Peer| {
+            trace!(
+                node_id,
+                ?peer,
+                "Received peer information from LocalSwarmDiscovery"
+            );
+
+            sender
+                .send(Message::Discovery(node_id.to_string(), peer.clone()))
+                .ok();
+        };
+
+        let mut addrs: HashMap<u16, Vec<IpAddr>> = HashMap::default();
+        let mut has_ipv4 = false;
+        let mut has_ipv6 = false;
+        for socketaddr in socketaddrs {
+            if !has_ipv6 && socketaddr.is_ipv6() {
+                has_ipv6 = true;
+            };
+            if !has_ipv4 && socketaddr.is_ipv4() {
+                has_ipv4 = true;
+            };
+            addrs
+                .entry(socketaddr.port())
+                .and_modify(|a| a.push(socketaddr.ip()))
+                .or_insert(vec![socketaddr.ip()]);
+        }
+
+        let ip_class = match (has_ipv4, has_ipv6) {
+            (true, true) => IpClass::V4AndV6,
+            (true, false) => IpClass::V4Only,
+            (false, true) => IpClass::V6Only,
+            // this case indicates no ip addresses were supplied, in which case, default to ipv4
+            (false, false) => IpClass::V4Only,
+        };
+        debug!(?addrs, ?ip_class, "creating discoverer");
+        let mut discoverer =
+            Discoverer::new_interactive(N0_LOCAL_SWARM.to_string(), node_id.to_string())
+                .with_callback(callback)
+                .with_ip_class(ip_class);
+        for addr in addrs {
+            discoverer = discoverer.with_addrs(addr.0, addr.1);
+        }
+        discoverer.spawn(rt)
+    }
+}
+
+impl From<&Peer> for DiscoveryItem {
+    fn from(peer_info: &Peer) -> Self {
+        let direct_addresses: BTreeSet<SocketAddr> = peer_info
+            .addrs()
+            .iter()
+            .map(|(ip, port)| SocketAddr::new(*ip, *port))
+            .collect();
+        DiscoveryItem {
+            provenance: PROVENANCE,
+            last_updated: None,
+            addr_info: AddrInfo {
+                relay_url: None,
+                direct_addresses,
+            },
+        }
+    }
+}
+
+impl Discovery for LocalSwarmDiscovery {
+    fn resolve(&self, _ep: Endpoint, node_id: NodeId) -> Option<BoxStream<Result<DiscoveryItem>>> {
+        let (send, recv) = flume::bounded(20);
+        let discovery_sender = self.sender.clone();
+        tokio::spawn(async move {
+            discovery_sender
+                .send_async(Message::SendAddrs(node_id, send))
+                .await
+                .ok();
+        });
+        Some(recv.into_stream().boxed())
```

Comment:
I would like the `tokio::spawn` side effect to be captured by the return value.

The idea being: If you call `.resolve()`, but then drop the returned stream, I would expect it to never actually kick off any work.

```suggestion
        let stream = async move {
            discovery_sender
                .send_async(Message::SendAddrs(node_id, send))
                .await?;
            Ok(recv.into_stream())
        };
        Some(stream.try_flatten_stream().boxed())
```

This way, the `send_async` is run within the returned stream. Due to the way that `try_flatten_stream` works, this also means it will surface any errors that happen from `send_async`.

Reply:
this is very elegant i'm into it!

Reply:
ahhh, except it relies on a trait from the `futures` crate, and we have been purging our codebase of that dependency.